refactor!: change suda#system to a list-based api

aarondill opened the pull request:
This eliminates the need for shell-escaping, and removes possibility of accidental shell-injection.

This is technically a breaking change, since suda#system is globally exposed; however this is *not* a documented public api, so the chance of breakage is minimal.

Note that this *changes the behavior of g:suda#executeable! It is no longer shell-processed, but rather treated as a path/name to an executable and processed by the os in that way.
Should we split suda#executable on spaces to preserve the old behavior? Should we allow suda#executable to (optionally) be a list to allow passing arguments?


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **Refactor**
	- Enhanced the internal handling of command execution and options in the `suda` functionality for improved performance and flexibility.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

aarondill committed:
refactor!: change suda#system to a list-based api

This eliminates the need for shell-escaping, and removes possibility of accidental shell-injection.

This is technically a breaking change, since suda#system is globally exposed; however this is *not* a documented public api, so the chance of breakage is minimal.

Note that this *changes the behavior of g:suda#executeable! It is no longer shell-processed, but rather treated as a path/name to an executable and processed by the os in that way.
Should we split suda#executable on spaces to preserve the old behavior? Should we allow suda#executable to (optionally) be a list to allow passing arguments?

aarondill commented:
@lambdalisue what do you think about this?

lambdalisue commented:
Sorry for late. Well, probably we should add `suda#systemlist` function and use it internally.

aarondill commented:
So you're saying have both `suda#system` and `suda#systemlist`? I was thinking about this, but went with the current approach to reduce code duplication. since the two functions do nearly the same thing, except calling `system` and `systemlist` respectively.
if you think it would increase clarity, I could implement separate functions as wrappers around the current `suda#system` implementation.

lambdalisue commented:
> if you think it would increase clarity, I could implement separate functions as wrappers around the current suda#system implementation.

Yes please. We may remove `suda#system` later but not now 👍

aarondill commented:
@lambdalisue to be clear, I was thinking about the following simple implementation 
```vim
" name tbd
function s:_system(cmd, list, ...)
... " the current implementation
endfunction

function suda#system(cmd, ...)
return s:_system(cmd, 0, ...)
endfunction

function suda#systemlist(cmd, ...)
return s:_system(cmd, 1, ...)
endfunction
```
is something like this acceptable?

lambdalisue commented:
I'm sorry. I meant something like

```vim
" arguments must like systemlist()
function! suda#systemlist(cmd, ...) abort
  " ...
endfunction

" arguments must like sysmte()
function! suda#system(cmd, ...) abort
  " ...
  return suda#systemlist(...)
endfunction
```

So that we can simply remove `suda#system` later.

aarondill commented:
that wouldn't work the way you're thinking.
`suda#system` mimics the vim  `system` function.
`system` and `systemlist` have the same arguments (`cmd` as a string or list, and `input` as optional string/list/buffer id)
they differ in return types. `system` returns a string representing the command's output, while `systemlist` returns a list representing the command's output (split by newlines, preserving nulls as embedded newlines)

this is why I went the approach that i did, the input arguments are the same, but the return type varies depending on if it calls `system` or `systemlist`

lambdalisue commented:
Apologies. I completely forgot that I had previously made the exact same mistake a long time ago. Nevertheless, in this scenario, could we consider the following approach?

```vim
function! suda#systemlist(cmd, ...) abort
  " ...
endfunction

function! suda#system(cmd, ...) abort
  let result = call('suda#systemlist', [a:cmd] + a:000)
  return join(result, "\n")
endfunction
```

Alternatively, if you find `suda#system()` to be too redundant, we can omit it and perhaps introduce `suda#systemlist()` as part of a new major release.

aarondill commented:
@lambdalisue This aproach also fails because the output of `systemlist` translates NULs into NLs. If we only had to spawn the process once, we could have a helper which returns the argument list, but we spawn the process once, check for failure and potentially spawn it again, meaning we would have to pass in a funcref anyways.
I don't see a good way to implement it without passing a funcref to `system`/`systemlist` to a helper, reimplementing `system`'s processing of the output (which is already provided by VimScript :pensive:), or a lot of code duplication.

> systemlist({expr} [, {input}])				systemlist()
>
> Same [as](https://vimhelp.org/motion.txt.html#as) [system()](https://vimhelp.org/builtin.txt.html#system%28%29)[,](https://vimhelp.org/motion.txt.html#%2C) but returns [a](https://vimhelp.org/insert.txt.html#a) [List](https://vimhelp.org/eval.txt.html#List) with lines (parts of
		output separated by NL) with NULs transformed into NLs. Output
		[is](https://vimhelp.org/motion.txt.html#is) the same [as](https://vimhelp.org/motion.txt.html#as) [readfile()](https://vimhelp.org/builtin.txt.html#readfile%28%29) will output with {binary} argument
		set to "[b](https://vimhelp.org/motion.txt.html#b)"[,](https://vimhelp.org/motion.txt.html#%2C) except that there [is](https://vimhelp.org/motion.txt.html#is) no extra empty item when the
		result ends in [a](https://vimhelp.org/insert.txt.html#a) NL.
		Note that on [MS-Windows](https://vimhelp.org/os_win32.txt.html#MS-Windows) you may get trailing CR characters.

lambdalisue commented:
@aarondill Well, nvm for NULs while Vim's `system()` doesn't support anyway. Using `join(systemlist(...), "\n")` should be enough I think.

aarondill commented:
@lambdalisue that approach won't work unless we remove the NULs before joining (since they're represented by newlines).

it seems that the default `system` turns NULs into \u0001 so they don't break strings. should we do this too, or just remove them from the string?

aarondill commented:
> To make the result more system-independent, the shell output
		[is](https://vimhelp.org/motion.txt.html#is) filtered to replace [<CR>](https://vimhelp.org/motion.txt.html#%3CCR%3E) with [<NL>](https://vimhelp.org/motion.txt.html#%3CNL%3E) for Macintosh, and
		[<CR>](https://vimhelp.org/motion.txt.html#%3CCR%3E)[<NL>](https://vimhelp.org/motion.txt.html#%3CNL%3E) with [<NL>](https://vimhelp.org/motion.txt.html#%3CNL%3E) for DOS-like systems.
		To avoid the [string](https://vimhelp.org/eval.txt.html#string) being truncated [at](https://vimhelp.org/motion.txt.html#at) [a](https://vimhelp.org/insert.txt.html#a) NUL, all NUL
		characters are replaced with SOH (0x01).

lambdalisue commented:
> it seems that the default system turns NULs into \u0001 so they don't break strings. should we do this too, or just remove them from the string?

Let's follow Vim's behavior 👍

aarondill committed:
refactor: implement suda#systemlist

This replaces the 3 argument suda#system implementation and follows the vimscript implementation.

aarondill commented:
@lambdalisue I've implemented this change in a583f3a7e764daee6f119887c879f8b048fc80fa

aarondill commented on autoload/suda.vim:
```diff
@@ -28,24 +36,34 @@ function! suda#system(cmd, ...) abort
+    let cmd = s:get_command(['-p', '', '-S'], a:cmd)
   endif
-  return system(cmd, password . "\n" . (a:0 ? a:1 : ''))
+  return systemlist(cmd, password . "\n" . (a:0 ? a:1 : ''))
```
1. passing through stdin is about as safe as it gets (more so than env vars)
2. this is already the implementation and not added in this commit.

aarondill committed:
fix: #system passes empty list to systemlist

use call() instead to ensure that it passes the right amount of args

lambdalisue commented on autoload/suda.vim:
```diff
@@ -57,6 +75,7 @@ function! suda#read(expr, ...) abort range
         \)
 
   if filereadable(path)
+    " TODO: (aarondill) can we use readfile() here?
```
Unfortunately, no. `readfile()` does not support `++ff=FF` or similar.

lambdalisue commented on autoload/suda.vim:
```diff
@@ -110,6 +127,7 @@ function! suda#write(expr, ...) abort range
   let tempfile = tempname()
   try
     let path_exists = !empty(getftype(path))
+    " TODO: (aarondill) can we use writefile() here?
```
Unfortunately, no. Same as `readfile()`.

lambdalisue commented:
LGTM
